File: web/src/generator/expression.ts
import { Identifier, RuleDeclaration, Node } from "./node";

export class Expression extends Node {
	walk(f: (expr: Expression) => Expression): Expression {
		return f(this);
	}
	eq(_other: Expression): boolean {
		return false;
	}
	prec!: number;
}

Expression.prototype.prec = 10;

export class NameExpression extends Expression {
	readonly id: Identifier;
	readonly args: readonly Expression[];

	constructor(start: number, id: Identifier, args: readonly Expression[]) {
		super(start);
		this.id = id;
		this.args = args;
	}
	toString(): string {
		return this.id.name + (this.args.length ? `<${this.args.join()}>` : "");
	}
	eq(other: NameExpression): boolean {
		return this.id.name == other.id.name && exprsEq(this.args, other.args);
	}
	walk(f: (expr: Expression) => Expression): Expression {
		let args = walkExprs(this.args, f);
		return f(args == this.args ? this : new NameExpression(this.start, this.id, args));
	}
}

export class SpecializeExpression extends Expression {
	readonly type: string;
	readonly props: readonly Prop[];
	readonly token: Expression;
	readonly content: Expression;

	constructor(
		start: number,
		type: string,
		props: readonly Prop[],
		token: Expression,
		content: Expression,
	) {
		super(start);
		this.type = type;
		this.props = props;
		this.token = token;
		this.content = content;
	}
	toString() {
		return `@${this.type}[${this.props.join(",")}]<${this.token}, ${this.content}>`;
	}
	eq(other: SpecializeExpression): boolean {
		return (
			this.type == other.type &&
			Prop.eqProps(this.props, other.props) &&
			exprEq(this.token, other.token) &&
			exprEq(this.content, other.content)
		);
	}
	walk(f: (expr: Expression) => Expression): Expression {
		let token = this.token.walk(f),
			content = this.content.walk(f);
		return f(
			token == this.token && content == this.content
				? this
				: new SpecializeExpression(this.start, this.type, this.props, token, content),
		);
	}
}

export class InlineRuleExpression extends Expression {
	readonly rule: RuleDeclaration;

	constructor(start: number, rule: RuleDeclaration) {
		super(start);
		this.rule = rule;
	}

	toString() {
		let rule = this.rule;
		return `${rule.id}${rule.props.length ? `[${rule.props.join(",")}]` : ""} { ${rule.expr} }`;
	}
	eq(other: InlineRuleExpression): boolean {
		let rule = this.rule,
			oRule = other.rule;
		return (
			exprEq(rule.expr, oRule.expr) &&
			rule.id.name == oRule.id.name &&
			Prop.eqProps(rule.props, oRule.props)
		);
	}
	walk(f: (expr: Expression) => Expression): Expression {
		let rule = this.rule,
			expr = rule.expr.walk(f);
		return f(
			expr == rule.expr
				? this
				: new InlineRuleExpression(
						this.start,
						new RuleDeclaration(rule.start, rule.id, rule.props, [], expr),
					),
		);
	}
}

export class ChoiceExpression extends Expression {
	readonly exprs: readonly Expression[];

	constructor(start: number, exprs: readonly Expression[]) {
		super(start);
		this.exprs = exprs;
	}
	toString(): string {
		return this.exprs.map((e) => maybeParens(e, this)).join(" | ");
	}
	eq(other: ChoiceExpression): boolean {
		return exprsEq(this.exprs, other.exprs);
	}
	walk(f: (expr: Expression) => Expression): Expression {
		let exprs = walkExprs(this.exprs, f);
		return f(exprs == this.exprs ? this : new ChoiceExpression(this.start, exprs));
	}
}

ChoiceExpression.prototype.prec = 1;

export class SequenceExpression extends Expression {
	readonly exprs: readonly Expression[];
	readonly markers: readonly (readonly ConflictMarker[])[];
	readonly empty: boolean;

	constructor(
		start: number,
		exprs: readonly Expression[],
		markers: readonly (readonly ConflictMarker[])[],
		empty = false,
	) {
		super(start);
		this.exprs = exprs;
		this.markers = markers;
		this.empty = empty;
	}
	toString(): string {
		return this.empty ? "()" : this.exprs.map((e) => maybeParens(e, this)).join(" ");
	}
	eq(other: SequenceExpression): boolean {
		return (
			exprsEq(this.exprs, other.exprs) &&
			this.markers.every((m, i) => {
				let om = other.markers[i];
				return m.length == om.length && m.every((x, i) => x.eq(om[i]));
			})
		);
	}
	walk(f: (expr: Expression) => Expression): Expression {
		let exprs = walkExprs(this.exprs, f);
		return f(
			exprs == this.exprs
				? this
				: new SequenceExpression(this.start, exprs, this.markers, this.empty && !exprs.length),
		);
	}
}

SequenceExpression.prototype.prec = 2;

export class ConflictMarker extends Node {
	readonly id: Identifier;
	readonly type: "ambig" | "prec";

	constructor(start: number, id: Identifier, type: "ambig" | "prec") {
		super(start);
		this.id = id;
		this.type = type;
	}

	toString(): string {
		return (this.type == "ambig" ? "~" : "!") + this.id.name;
	}

	eq(other: ConflictMarker): boolean {
		return this.id.name == other.id.name && this.type == other.type;
	}
}

export class RepeatExpression extends Expression {
	readonly expr: Expression;
	readonly kind: "?" | "*" | "+";

	constructor(start: number, expr: Expression, kind: "?" | "*" | "+") {
		super(start);
		this.expr = expr;
		this.kind = kind;
	}
	toString(): string {
		return maybeParens(this.expr, this) + this.kind;
	}
	eq(other: RepeatExpression): boolean {
		return exprEq(this.expr, other.expr) && this.kind == other.kind;
	}
	walk(f: (expr: Expression) => Expression): Expression {
		let expr: Expression = this.expr.walk(f);
		return f(expr == this.expr ? this : new RepeatExpression(this.start, expr, this.kind));
	}
}

RepeatExpression.prototype.prec = 3;

export class LiteralExpression extends Expression {
	readonly value: string;

	constructor(start: number, value: string) {
		super(start);
		this.value = value;
	}
	toString(): string {
		return JSON.stringify(this.value);
	}
	eq(other: LiteralExpression): boolean {
		return this.value == other.value;
	}
}

export class SetExpression extends Expression {
	readonly ranges: [number, number][];
	readonly inverted: boolean;

	constructor(start: number, ranges: [number, number][], inverted: boolean) {
		super(start);
		this.ranges = ranges;
		this.inverted = inverted;
	}
	toString() {
		return `[${this.inverted ? "^" : ""}${this.ranges.map(([a, b]) => {
			return String.fromCodePoint(a) + (b == a + 1 ? "" : "-" + String.fromCodePoint(b));
		})}]`;
	}
	eq(other: SetExpression): boolean {
		return (
			this.inverted == other.inverted &&
			this.ranges.length == other.ranges.length &&
			this.ranges.every(([a, b], i) => {
				let [x, y] = other.ranges[i];
				return a == x && b == y;
			})
		);
	}
}

export class AnyExpression extends Expression {
	constructor(start: number) {
		super(start);
	}
	toString() {
		return "_";
	}
	eq() {
		return true;
	}
}

function walkExprs(
	exprs: readonly Expression[],
	f: (expr: Expression) => Expression,
): readonly Expression[] {
	let result: Expression[] | null = null;
	for (let i = 0; i < exprs.length; i++) {
		let expr = exprs[i].walk(f);
		if (expr != exprs[i] && !result) result = exprs.slice(0, i);
		if (result) result.push(expr);
	}
	return result || exprs;
}

export const CharClasses: { [name: string]: [number, number][] } = {
	asciiLetter: [
		[65, 91],
		[97, 123],
	],
	asciiLowercase: [[97, 123]],
	asciiUppercase: [[65, 91]],
	digit: [[48, 58]],
	whitespace: [
		[9, 14],
		[32, 33],
		[133, 134],
		[160, 161],
		[5760, 5761],
		[8192, 8203],
		[8232, 8234],
		[8239, 8240],
		[8287, 8288],
		[12288, 12289],
	],
	eof: [[0xffff, 0xffff]],
};

export class CharClass extends Expression {
	readonly type: string;

	constructor(start: number, type: string) {
		super(start);
		this.type = type;
	}
	toString(): string {
		return "@" + this.type;
	}
	eq(expr: CharClass): boolean {
		return this.type == expr.type;
	}
}

export function exprEq(a: Expression, b: Expression): boolean {
	return a.constructor == b.constructor && a.eq(b as any);
}

export function exprsEq(a: readonly Expression[], b: readonly Expression[]): boolean {
	return a.length == b.length && a.every((e, i) => exprEq(e, b[i]));
}

export class Prop extends Node {
	readonly at: boolean;
	readonly name: string;
	readonly value: readonly PropPart[];

	constructor(start: number, at: boolean, name: string, value: readonly PropPart[]) {
		super(start);
		this.at = at;
		this.name = name;
		this.value = value;
	}

	eq(other: Prop): boolean {
		return (
			this.name == other.name &&
			this.value.length == other.value.length &&
			this.value.every((v, i) => v.value == other.value[i].value && v.name == other.value[i].name)
		);
	}

	toString(): string {
		let result = (this.at ? "@" : "") + this.name;
		if (this.value.length) {
			result += "=";
			for (let { name, value } of this.value)
				result += name ? `{${name}}` : /[^\w-]/.test(value!) ? JSON.stringify(value) : value;
		}
		return result;
	}

	static eqProps(a: readonly Prop[], b: readonly Prop[]): boolean {
		return a.length == b.length && a.every((p, i) => p.eq(b[i]));
	}
}

export class PropPart extends Node {
	readonly value: string | null;
	readonly name: string | null;

	constructor(start: number, value: string | null, name: string | null) {
		super(start);
		this.value = value;
		this.name = name;
	}
}

function maybeParens(node: Expression, parent: Expression) {
	return node.prec < parent.prec ? "(" + node.toString() + ")" : node.toString();
}

File: dotnet/Rezel/Generator/Expression.cs
using System.Text;
using System.Text.Json;

namespace Rezel.Generator;

public abstract class Expression : Node
{
    protected Expression(int start) : base(start) { }

    public virtual int Prec => 10;

    public virtual Expression Walk(Func<Expression, Expression> f) => f(this);

    public virtual bool Eq(Expression other) => false;

    public static readonly Dictionary<string, int[][]> CharClasses = new()
    {
        ["asciiLetter"] = new int[][] { new int[] { 65, 91 }, new int[] { 97, 123 } },
        ["asciiLowercase"] = new int[][] { new int[] { 97, 123 } },
        ["asciiUppercase"] = new int[][] { new int[] { 65, 91 } },
        ["digit"] = new int[][] { new int[] { 48, 58 } },
        ["whitespace"] = new int[][]
        {
            new int[] { 9, 14 },
            new int[] { 32, 33 },
            new int[] { 133, 134 },
            new int[] { 160, 161 },
            new int[] { 5760, 5761 },
            new int[] { 8192, 8203 },
            new int[] { 8232, 8234 },
            new int[] { 8239, 8240 },
            new int[] { 8287, 8288 },
            new int[] { 12288, 12289 },
        },
        ["eof"] = new int[][] { new int[] { 0xffff, 0xffff } },
    };

    public static bool ExprEq(Expression a, Expression b)
    {
        return a.GetType() == b.GetType() && a.Eq(b);
    }

    public static bool ExprsEq(Expression[] a, Expression[] b)
    {
        if (a.Length != b.Length) return false;
        for (int i = 0; i < a.Length; i++)
            if (!ExprEq(a[i], b[i])) return false;
        return true;
    }

    internal static Expression[] WalkExprs(Expression[] exprs, Func<Expression, Expression> f)
    {
        List<Expression>? result = null;
        for (int i = 0; i < exprs.Length; i++)
        {
            var expr = exprs[i].Walk(f);
            if (expr != exprs[i] && result == null)
                result = new List<Expression>(exprs[..i]);
            if (result != null)
                result.Add(expr);
        }
        return result?.ToArray() ?? exprs;
    }

    internal static string MaybeParens(Expression node, Expression parent)
    {
        return node.Prec < parent.Prec ? $"({node})" : node.ToString()!;
    }
}

public sealed class NameExpression : Expression
{
    public readonly Identifier Id;
    public readonly Expression[] Args;

    public NameExpression(int start, Identifier id, Expression[] args) : base(start)
    {
        Id = id;
        Args = args;
    }

    public override string ToString()
    {
        if (Args.Length == 0) return Id.Name;
        return $"{Id.Name}<{string.Join(",", Args.Select(a => a.ToString()))}>";
    }

    public override bool Eq(Expression other)
    {
        var o = (NameExpression)other;
        return Id.Name == o.Id.Name && ExprsEq(Args, o.Args);
    }

    public override Expression Walk(Func<Expression, Expression> f)
    {
        var args = WalkExprs(Args, f);
        return f(args == Args ? this : new NameExpression(Start, Id, args));
    }
}

public sealed class SpecializeExpression : Expression
{
    public readonly string Type;
    public readonly Prop[] Props;
    public readonly Expression Token;
    public readonly Expression Content;

    public SpecializeExpression(int start, string type, Prop[] props, Expression token, Expression content) : base(start)
    {
        Type = type;
        Props = props;
        Token = token;
        Content = content;
    }

    public override string ToString()
    {
        return $"@{Type}[{string.Join(",", Props.Select(p => p.ToString()))}]<{Token}, {Content}>";
    }

    public override bool Eq(Expression other)
    {
        var o = (SpecializeExpression)other;
        return Type == o.Type && Prop.EqProps(Props, o.Props) && ExprEq(Token, o.Token) && ExprEq(Content, o.Content);
    }

    public override Expression Walk(Func<Expression, Expression> f)
    {
        var token = Token.Walk(f);
        var content = Content.Walk(f);
        return f(
            token == Token && content == Content
                ? this
                : new SpecializeExpression(Start, Type, Props, token, content)
        );
    }
}

public sealed class InlineRuleExpression : Expression
{
    public readonly RuleDeclaration Rule;

    public InlineRuleExpression(int start, RuleDeclaration rule) : base(start)
    {
        Rule = rule;
    }

    public override string ToString()
    {
        var rule = Rule;
        var propsPart = rule.Props.Length > 0 ? $"[{string.Join(",", rule.Props.Select(p => p.ToString()))}]" : "";
        return $"{rule.Id.Name}{propsPart} {{ {rule.Expr} }}";
    }

    public override bool Eq(Expression other)
    {
        var rule = Rule;
        var oRule = ((InlineRuleExpression)other).Rule;
        return ExprEq(rule.Expr, oRule.Expr) && rule.Id.Name == oRule.Id.Name && Prop.EqProps(rule.Props, oRule.Props);
    }

    public override Expression Walk(Func<Expression, Expression> f)
    {
        var rule = Rule;
        var expr = rule.Expr.Walk(f);
        return f(
            expr == rule.Expr
                ? this
                : new InlineRuleExpression(Start, new RuleDeclaration(rule.Start, rule.Id, rule.Props, rule.Params, expr))
        );
    }
}

public sealed class ChoiceExpression : Expression
{
    public override int Prec => 1;

    public readonly Expression[] Exprs;

    public ChoiceExpression(int start, Expression[] exprs) : base(start)
    {
        Exprs = exprs;
    }

    public override string ToString()
    {
        return string.Join(" | ", Exprs.Select(e => MaybeParens(e, this)));
    }

    public override bool Eq(Expression other)
    {
        return ExprsEq(Exprs, ((ChoiceExpression)other).Exprs);
    }

    public override Expression Walk(Func<Expression, Expression> f)
    {
        var exprs = WalkExprs(Exprs, f);
        return f(exprs == Exprs ? this : new ChoiceExpression(Start, exprs));
    }
}

public sealed class SequenceExpression : Expression
{
    public override int Prec => 2;

    public readonly Expression[] Exprs;
    public readonly ConflictMarker[][] Markers;
    public readonly bool Empty;

    public SequenceExpression(int start, Expression[] exprs, ConflictMarker[][] markers, bool empty = false) : base(start)
    {
        Exprs = exprs;
        Markers = markers;
        Empty = empty;
    }

    public override string ToString()
    {
        if (Empty) return "()";
        return string.Join(" ", Exprs.Select(e => MaybeParens(e, this)));
    }

    public override bool Eq(Expression other)
    {
        var o = (SequenceExpression)other;
        if (!ExprsEq(Exprs, o.Exprs)) return false;
        for (int i = 0; i < Markers.Length; i++)
        {
            var m = Markers[i];
            var om = o.Markers[i];
            if (m.Length != om.Length) return false;
            for (int j = 0; j < m.Length; j++)
                if (!m[j].Eq(om[j])) return false;
        }
        return true;
    }

    public override Expression Walk(Func<Expression, Expression> f)
    {
        var exprs = WalkExprs(Exprs, f);
        return f(
            exprs == Exprs
                ? this
                : new SequenceExpression(Start, exprs, Markers, Empty && exprs.Length == 0)
        );
    }
}

public sealed class ConflictMarker : Node
{
    public readonly Identifier Id;
    public readonly string Type;

    public ConflictMarker(int start, Identifier id, string type) : base(start)
    {
        Id = id;
        Type = type;
    }

    public override string ToString()
    {
        return (Type == "ambig" ? "~" : "!") + Id.Name;
    }

    public bool Eq(ConflictMarker other)
    {
        return Id.Name == other.Id.Name && Type == other.Type;
    }
}

public sealed class RepeatExpression : Expression
{
    public override int Prec => 3;

    public readonly Expression Expr;
    public readonly string Kind;

    public RepeatExpression(int start, Expression expr, string kind) : base(start)
    {
        Expr = expr;
        Kind = kind;
    }

    public override string ToString()
    {
        return MaybeParens(Expr, this) + Kind;
    }

    public override bool Eq(Expression other)
    {
        var o = (RepeatExpression)other;
        return ExprEq(Expr, o.Expr) && Kind == o.Kind;
    }

    public override Expression Walk(Func<Expression, Expression> f)
    {
        var expr = Expr.Walk(f);
        return f(expr == Expr ? this : new RepeatExpression(Start, expr, Kind));
    }
}

public sealed class LiteralExpression : Expression
{
    public readonly string Value;

    public LiteralExpression(int start, string value) : base(start)
    {
        Value = value;
    }

    public override string ToString()
    {
        return JsonSerializer.Serialize(Value);
    }

    public override bool Eq(Expression other)
    {
        return Value == ((LiteralExpression)other).Value;
    }
}

public sealed class SetExpression : Expression
{
    public readonly int[][] Ranges;
    public readonly bool Inverted;

    public SetExpression(int start, int[][] ranges, bool inverted) : base(start)
    {
        Ranges = ranges;
        Inverted = inverted;
    }

    public override string ToString()
    {
        var sb = new StringBuilder("[");
        if (Inverted) sb.Append('^');
        foreach (var range in Ranges)
        {
            sb.Append(char.ConvertFromUtf32(range[0]));
            if (range[1] != range[0] + 1)
            {
                sb.Append('-');
                sb.Append(char.ConvertFromUtf32(range[1]));
            }
        }
        sb.Append(']');
        return sb.ToString();
    }

    public override bool Eq(Expression other)
    {
        var o = (SetExpression)other;
        if (Inverted != o.Inverted || Ranges.Length != o.Ranges.Length) return false;
        for (int i = 0; i < Ranges.Length; i++)
        {
            if (Ranges[i][0] != o.Ranges[i][0] || Ranges[i][1] != o.Ranges[i][1]) return false;
        }
        return true;
    }
}

public sealed class AnyExpression : Expression
{
    public AnyExpression(int start) : base(start) { }

    public override string ToString() => "_";

    public override bool Eq(Expression other) => true;
}

public sealed class CharClass : Expression
{
    public readonly string Type;

    public CharClass(int start, string type) : base(start)
    {
        Type = type;
    }

    public override string ToString() => "@" + Type;

    public override bool Eq(Expression other)
    {
        return Type == ((CharClass)other).Type;
    }
}

public sealed class Prop : Node
{
    public readonly bool At;
    public readonly string Name;
    public readonly PropPart[] Value;

    public Prop(int start, bool at, string name, PropPart[] value) : base(start)
    {
        At = at;
        Name = name;
        Value = value;
    }

    public override string ToString()
    {
        var result = new StringBuilder();
        if (At) result.Append('@');
        result.Append(Name);
        if (Value.Length > 0)
        {
            result.Append('=');
            foreach (var part in Value)
                result.Append(part.Name != null ? $"{{{part.Name}}}" : NeedsQuoting(part.Value) ? JsonSerializer.Serialize(part.Value!) : part.Value);
        }
        return result.ToString();
    }

    public bool Eq(Prop other)
    {
        if (Name != other.Name || Value.Length != other.Value.Length) return false;
        for (int i = 0; i < Value.Length; i++)
            if (Value[i].Value != other.Value[i].Value || Value[i].Name != other.Value[i].Name) return false;
        return true;
    }

    public static bool EqProps(Prop[] a, Prop[] b)
    {
        if (a.Length != b.Length) return false;
        for (int i = 0; i < a.Length; i++)
            if (!a[i].Eq(b[i])) return false;
        return true;
    }

    private static bool NeedsQuoting(string? value)
    {
        if (value == null) return false;
        foreach (char c in value)
            if (!char.IsLetterOrDigit(c) && c != '_' && c != '-') return true;
        return false;
    }
}

public sealed class PropPart : Node
{
    public readonly string? Value;
    public readonly string? Name;

    public PropPart(int start, string? value, string? name) : base(start)
    {
        Value = value;
        Name = name;
    }
}
